### PR TITLE
Markdown: allow to end URL with balanced parenthesis

### PR DESCRIPTION
**Pull request by a-mr**
Mostly fixes bug 24 from #17340 (except escaping).

cc @timotheecour

### CODE REVIEW

File: lib/packages/docutils/rst.nim [hunk 1]
```diff
@@ -1246,15 +1246,44 @@ proc isUrl(p: RstParser, i: int): bool =
...
+  ## Returns `true` iff `token` is a closing parenthesis for some
+  ## previous opening parenthesis saved in `parensStack`.
+  ## This is according Markdown balanced parentheses rule
+  ## (https://spec.commonmark.org/0.29/#link-destination)
+  ## to allow links like
+  ## https://en.wikipedia.org/wiki/APL_(programming_language),
+  ## we use it for RST also.
+  result = false
+  if token.kind == tkPunct:
+    let sym = token.symbol
```
Comment (timotheecour) [inline lib/packages/docutils/rst.nim]:
```suggestion
    let c = token.symbol[0]
```
you only use c below

File: lib/packages/docutils/rst.nim [hunk 2]
```diff
@@ -1246,15 +1246,44 @@ proc isUrl(p: RstParser, i: int): bool =
     p.tok[i+3].kind == tkWord and
     p.tok[i].symbol in ["http", "https", "ftp", "telnet", "file"]
 
+proc checkParen(token: Token, parensStack: var seq[char]): bool {.inline.} =
```
Comment (timotheecour) [inline lib/packages/docutils/rst.nim:1249]: completely optional: can you split this into a more general reusable proc and then call it, and then in future work we can move this reusable proc to strbasics or strutils? (similar to https://dlang.org/library/std/algorithm/searching/balanced_parens.html, and same functionality as what you have)
```nim
proc balancedParens(a: openArray[char], parensStack: var seq[char]): bool =
  # xxx in future work, move to stdlib
proc checkParen(token: Token, parensStack: var seq[char]): bool {.inline.} =
  call balancedParens somehow
```

Comment (a-mr) [reply lib/packages/docutils/rst.nim:1249]: The idea of `balancedParens` is very different: it check whether parens are balanced or not.

`checkParen` does not always forbid unbalanced parens: e.g. `([)` and `(])` are legal, and `checkParen` will return 3 values `False,False,True` in both cases for 3 tokens.

File: lib/packages/docutils/rst.nim [hunk 3]
```diff
@@ -1246,15 +1246,44 @@ proc isUrl(p: RstParser, i: int): bool =
...
+  ## to allow links like
+  ## https://en.wikipedia.org/wiki/APL_(programming_language),
+  ## we use it for RST also.
+  result = false
+  if token.kind == tkPunct:
+    let c = token.symbol[0]
+    if c in {'(', '[', '{'}:  # push
+      parensStack.add c
+    elif c in {')', ']', '}'}:  # try pop
+      if parensStack.len > 0:
```
Comment (Araq) [inline lib/packages/docutils/rst.nim]: Pointless `if` statement, if the collection is empty the loop will iterate 0 times.

Comment (a-mr) [reply lib/packages/docutils/rst.nim]: removed, thanks.

**Pull request merged**